### PR TITLE
Improve lookup when we call a function from TClingCallFunc.

### PR DESCRIPTION
**Pull request by Teemperor**
It turns out that just calling every function like this
```
functionname[<template args>](args)
```
doesn't always result in correct code.

In a concrete example, PyROOT is currently not able to generate a
valid wrapper for comparing libc++ std::vector<int>::iterator objects.
The wrapper becomes invalid by over-specifying the template arguments
and trying to call:
```
std::__1::operator==<int *, int *>(a, b)
```

Now we first try to cast such functions to the correct type before
we start doing a lookup, e.g. above example would be:

```
((bool (&)(const std::__wrap_iter<int*>&,const std::__wrap_iter<int*>&))std::__1::operator==<int*, int*>)(i,i)

### CODE REVIEW

**Comment by pcanal** [comment]
Is the case where the function template argument are *not* determined by the argument still working?  I.e. Can we still invoke (after finding the right instantation):
```
   template <type T> const std::type_info GetTypeInfo() { return typeid(T); }
```
[This is 'just' an example of such functions.]

**Comment by vgvassilev** [comment]
@phsft-bot build!

File: core/metacling/src/TClingCallFunc.cxx [hunk 1]
```diff
@@ -387,16 +387,129 @@ void TClingCallFunc::make_narg_ctor(const unsigned N, ostringstream &typedefbuf,
...
+         make_arg(FirstArg.empty() ? 1U : 0U, typedefbuf, callbuf, class_name, indent_level); \
+      } \
+      else if (args == 1) { \
+         assert(Unary); \
+         callbuf << Spelling << "("; \
+         if (FirstArg.empty()) make_arg(0U, typedefbuf, callbuf, class_name, indent_level); \
+         else callbuf << FirstArg; \
+         callbuf << ")"; \
+      } \
+      else assert(false && "Overloaded operator should have either 1 or 2 arguments!"); \
```
Comment (vgvassilev) [inline core/metacling/src/TClingCallFunc.cxx]: What about llvm::unreachable?

File: core/metacling/src/TClingCallFunc.cxx [hunk 2]
```diff
@@ -387,16 +387,129 @@ void TClingCallFunc::make_narg_ctor(const unsigned N, ostringstream &typedefbuf,
...
+      break; \
+   }
+#include "clang/Basic/OperatorKinds.def"
+#undef OVERLOADED_OPERATOR
+#undef OVERLOADED_OPERATOR_MULTI
+   default: assert(false && "We don't handle OVERLOADED_OPERATOR_MULTI here.");
```
Comment (vgvassilev) [inline core/metacling/src/TClingCallFunc.cxx]: Likewise.

**Comment by pcanal** [comment]
> Also, when calling an operator, instead of using the function-style
way of doing "a->operator==(b)", we now just call "a == b"
because that seems to be correct way of calling overloaded operators.
Again, this should still resolve in the right lookup because we
provide the arguments with the exact type the target function
expects.

Note that the semantic of CallFunc is as follow:
* find (by whatever means) a **specific** function/overload 
* call that **specific** overload

Having the compilation of the wrapper function do **another** overload resolution defeats part of the purpose.

**Comment by pcanal** [comment]
>  the former way of calling this function isn't able to be resolved to
the correct function, 

This is *odd* ... the function was found (there is a decl for it) and thus should be callable.  Are we sure it is not 'just' the spelling of the name that is wrong?

**Comment by Axel-Naumann** [comment]
```
#include <iterator>
#include <vector>
std::__wrap_iter<int*> i;
bool b2 = ((bool (&)(const std::__wrap_iter<int*>&,const std::__wrap_iter<int*>&))std::__1::operator==<int*, int*>)(i,i);
```
That disables lookup, "dialing" exactly the function overload we want to invoke. And then invoking it.

@Teemperor - it would be *awesome* if you could fix CallFunc by using this approach. It will work in "all cases" :-)

File: core/meta/inc/TInterpreter.h [hunk 3]
```diff
@@ -337,6 +337,8 @@ class TInterpreter : public TNamed {
    virtual void   CallFunc_SetFuncProto(CallFunc_t* func, ClassInfo_t* info, const char* method, const std::vector<TypeInfo_t*> &proto, Long_t* Offset, ROOT::EFunctionMatchMode mode = ROOT::kConversionMatch) const = 0;
    virtual void   CallFunc_SetFuncProto(CallFunc_t* func, ClassInfo_t* info, const char* method, const std::vector<TypeInfo_t*> &proto, bool objectIsConst, Long_t* Offset, ROOT::EFunctionMatchMode mode = ROOT::kConversionMatch) const = 0;
 
+   virtual std::string CallFunc_GenerateWrapper(CallFunc_t* func) const = 0;
```
Comment (vgvassilev) [inline core/meta/inc/TInterpreter.h]: What about using `CallFunc_PrintWrapper`?

File: core/metacling/src/TClingCallFunc.cxx [hunk 4]
```diff
@@ -528,6 +623,433 @@ void TClingCallFunc::make_narg_ctor_with_return(const unsigned N, const string &
    buf << "}\n";
 }
 
+int TClingCallFunc::generate_wrapper(std::string& wrapper_name, std::string& wrapper) {
+  const FunctionDecl *FD = GetDecl();
```
Comment (vgvassilev) [inline core/metacling/src/TClingCallFunc.cxx]: If this is the only class member that we need we could transform this routine into a public static one taking a `DeclId_t`. We will call it from the inside with `GetDecl()->getCanonicalDecl()` see `TCling::GetDeclId` (not really sure why it is not static, we could have reused it).

From the outside, we won't need to have a special method in `TInterpreter` and `TCling` but rely on the static interface.

**Comment by vgvassilev** [comment]
@Axel-Naumann,

It seems this won't still work:
```
struct X {} x;
template<typename T> T operator==(X, bool);
template<typename T> bool operator==(X, T);

bool k1 = x == true;
bool k2 = ((bool(&)(X,bool))::operator==<bool>)(x, true);
```
Courtesy of @zygoloid https://godbolt.org/g/k27s5o

**Comment by Axel-Naumann** [comment]
Right - we've been there before, and the only way to invoke a `Decl` reliably is to generate the appropriate AST instead of generating source code. What I meant is: "The above approach is more likely to pass roottest and existing usage patterns than this PR (which fails roottest)." Thanks for pointing out the test case again!

**Comment by vgvassilev** [comment]
It looks like we should continue ahead with what Raphael has started as the proposed solution doesn't handle such cases...

**Comment by pcanal** [comment]
> It looks like we should continue ahead with what Raphael has started 
Are you talking about the code in this PR (which can't work as is) or something else?

File: core/metacling/src/TClingCallFunc.cxx (hunk 4 above)
Comment (Teemperor) [inline core/metacling/src/TClingCallFunc.cxx]: We use fInterp, fNormCtxt and the Decl, so we actually use every single data member in this method :)

Comment (vgvassilev) [inline core/metacling/src/TClingCallFunc.cxx]: Bummer..

**Comment by Teemperor** [comment]
@phsft-bot build!

**Comment by Teemperor** [comment]
Ok, using Axel's suggestion for doing a hot-fix for this issue. It doesn't work for member functions, but it should fix the libc++ STL.

**Comment by pcanal** [comment]
> Ok, using Axel's suggestion for doing a hot-fix for this issue. It doesn't work for member functions, but it should fix the libc++ STL.

Since you can spell the type of function member and call them via a functions pointer:
```
    typedef void (TBranch::*ReadLeaves_t)(TBuffer &b);
```
The same pattern should also be applicable to member functions.

File: core/metacling/test/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,6 @@
+# The code in this directory is compiled with -fvisibility=hidden because
+# it is set in the parent directory. This causes warnings when we link against
+# gtest, so we have to unset this here.
+string (REPLACE "-fvisibility=hidden" "" CMAKE_CXX_FLAGS ${CMAKE_CXX_FLAGS})
```
Comment (vgvassilev) [inline core/metacling/test/CMakeLists.txt]: @Axel-Naumann is that what we want, fixing the test issue on OS X?

Comment (Axel-Naumann) [inline core/metacling/test/CMakeLists.txt]: Maybe... but really, you should not be able to call into libCling except for the exported interfaces.

File: core/metacling/test/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,6 @@
...
+
+ROOT_ADD_UNITTEST_DIR(Core Cling RIO)
```
Comment (Axel-Naumann) [inline core/metacling/test/CMakeLists.txt]: Do *not* link to Cling!

File: core/metacling/test/CMakeLists.txt (hunk 5 above)
Comment (Axel-Naumann) [inline core/metacling/test/CMakeLists.txt]: Not linking to libCling should take care of this... I.e. you don't need to filter that CXXFLAG away,

File: core/meta/inc/TInterpreter.h [hunk 7]
```diff
@@ -337,6 +337,7 @@ class TInterpreter : public TNamed {
    virtual void   CallFunc_SetFuncProto(CallFunc_t* func, ClassInfo_t* info, const char* method, const std::vector<TypeInfo_t*> &proto, Long_t* Offset, ROOT::EFunctionMatchMode mode = ROOT::kConversionMatch) const = 0;
    virtual void   CallFunc_SetFuncProto(CallFunc_t* func, ClassInfo_t* info, const char* method, const std::vector<TypeInfo_t*> &proto, bool objectIsConst, Long_t* Offset, ROOT::EFunctionMatchMode mode = ROOT::kConversionMatch) const = 0;
 
+   virtual std::string CallFunc_PrintWrapper(CallFunc_t *func) const = 0;
```
Comment (Axel-Naumann) [inline core/meta/inc/TInterpreter.h]: "Printing" sounds like the wrong verb here. Can we call it `GetWrapper()` or `GetWrapperCode()`

File: core/metacling/src/TCling.cxx [hunk 8]
```diff
@@ -6846,6 +6846,15 @@ void TCling::CallFunc_SetFuncProto(CallFunc_t* func, ClassInfo_t* info, const ch
    f->SetFuncProto(ci, method, funcProto, objectIsConst, offset, mode);
 }
 
+std::string TCling::CallFunc_GetWrapperCode(CallFunc_t *func) const
+{
+   TClingCallFunc *f = (TClingCallFunc *)func;
+   std::string wrapper_name;
+   std::string wrapper;
+   f->print_wrapper(wrapper_name, wrapper);
```
Comment (pcanal) [inline core/metacling/src/TCling.cxx]: Does print_wrapper actually show things on the terminal?  If not, wouldn't it be better to call it get_wrapper_code?

**Comment by pcanal** [comment]
Did you resolve the issue of member functions?

File: core/metacling/src/TCling.cxx (hunk 8 above)
Comment (Teemperor) [inline core/metacling/src/TCling.cxx]: Seems like this PR got already merged, so I renamed this in #642.

**Pull request merged**